src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -123,36 +123,53 @@ class JdbcReadJournal(config: Config, configPath: String)(implicit val system: E
   }
 
-  private def currentJournalEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[PersistentRepr, NotUsed] =
-    readJournalDao.messages(persistenceId, fromSequenceNr, toSequenceNr, Long.MaxValue)
+  private def currentJournalEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long, max: Long): Source[PersistentRepr, NotUsed] =
+    readJournalDao.messages(persistenceId, fromSequenceNr, toSequenceNr, max)
       .mapAsync(1)(deserializedRepr => Future.fromTry(deserializedRepr))
 
   override def currentEventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] =
-    currentJournalEventsByPersistenceId(persistenceId, fromSequenceNr, toSequenceNr)
+    currentJournalEventsByPersistenceId(persistenceId, fromSequenceNr, toSequenceNr, Long.MaxValue)
       .mapConcat(adaptEvents)
       .map(repr => EventEnvelope(Sequence(repr.sequenceNr), repr.persistenceId, repr.sequenceNr, repr.payload))
 
-  override def eventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] =
-    Source.unfoldAsync[Long, Seq[EventEnvelope]](Math.max(1, fromSequenceNr)) { (from: Long) =>
-      def nextFromSeqNr(xs: Seq[EventEnvelope]): Long = {
-        if (xs.isEmpty) from else xs.map(_.sequenceNr).max + 1
-      }
-      from match {
-        case x if x > toSequenceNr => Future.successful(None)
-        case _ =>
-          delaySource
-            .flatMapConcat { _ =>
-              currentJournalEventsByPersistenceId(persistenceId, from, toSequenceNr)
-                .take(readJournalConfig.maxBufferSize)
-            }
-            .mapConcat(adaptEvents)
-            .map(repr => EventEnvelope(Sequence(repr.sequenceNr), repr.persistenceId, repr.sequenceNr, repr.payload))
-            .runWith(Sink.seq).map { xs =>
-              val newFromSeqNr = nextFromSeqNr(xs)
-              Some((newFromSeqNr, xs))
+  override def eventsByPersistenceId(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long): Source[EventEnvelope, NotUsed] = {
+    import JdbcReadJournal._
+    val batchSize = readJournalConfig.maxBufferSize
+
+    Source.unfoldAsync[(Long, FlowControl), Seq[PersistentRepr]]((Math.max(1, fromSequenceNr), Continue)) {
+      case (from, control) =>
+        def retrieveNextBatch(): Future[Option[((Long, FlowControl), Seq[PersistentRepr])]] = {
+          for {
+            xs <- currentJournalEventsByPersistenceId(persistenceId, from, toSequenceNr, batchSize).runWith(Sink.seq)
+          } yield {
+            val hasMoreEvents = xs.size == batchSize
+            // Events are ordered by sequence number, therefore the last one is the largest
+            val hasLastEvent = xs.lastOption.exists(last => last.sequenceNr >= toSequenceNr)
+            val nextControl: FlowControl =
+              if (hasLastEvent || from > toSequenceNr) Stop
```

Comment:
If `hasLastEvent` is true, we should `ContinueDelayed`. This is expected to be an infinite stream of events. 

I think we will need another private method that we can use here in which we indicate if we want to stop on the last event or if we want to continue polling. I mean, `currentEventsByPersistenceId` should also benefit from this optimization. 

For `eventsByTag` we have `private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long])`. And we use `terminateAfterOffset` to decide if it should keep polling or not.

Reply:
The trait we are implementing see [EventsByPersistenceIdQuery.eventsByPersistenceId](https://doc.akka.io/api/akka/current/akka/persistence/query/scaladsl/EventsByPersistenceIdQuery.html#eventsByPersistenceId(persistenceId:String,fromSequenceNr:Long,toSequenceNr:Long):akka.stream.scaladsl.Source[akka.persistence.query.EventEnvelope,akka.NotUsed]) has the `toSequenceNr` as a mandatory parameter, it documents the following:
> You can retrieve a subset of all events by specifying fromSequenceNr and toSequenceNr or use 0L and Long.MaxValue respectively to retrieve all events.

I agree with the point that `currentEventsByPersistenceId` could also benefit from this optimization, this requires retrieving the current `toSequenceNr` first though.